text/0000-unsafe-extern-blocks.md

```diff
@@ -0,0 +1,120 @@
+
+# Summary
+[summary]: #summary
+
+In Edition 2024 it is `unsafe` to declare an `extern` function or static, but external functions and statics *can* be safe to use after the initial declaration.
+
+# Motivation
+[motivation]: #motivation
+
+Simply declaring extern items, even without ever using them, can cause Undefined Behavior.
+When performing cross-language compilation, attributes on one function declaration can flow to the foreign declaration elsewhere within LLVM and cause a miscompilation.
+In Rust we consider all sources of Undefined Behavior to be `unsafe`, and so we must make declaring extern blocks be `unsafe`.
+The up-side to this change is that in the new style it will be possible to declare an extern fn that's safe to call after the initial unsafe declaration.
+
+# Guide-level explanation
+[guide-level-explanation]: #guide-level-explanation
+
+Rust can utilize functions and statics from foreign code that are provided during linking, though it is `unsafe` to do so.
+
+An `extern` block can be placed anywhere a function declaration could appear (generally at the top level of a module).
+
+* On editions >= 2024, you *must* write all `extern` blocks as `unsafe extern`.
+* On editions < 2024, you *may* write `unsafe extern`, or you can write an `extern` block without the `unsafe` keyword. Writing an `extern` block without the `unsafe` keyword is provided for compatibility only, and will generate a warning.
+* `unsafe extern` interacts with the `unsafe_code` lint, and a `deny` or `forbid` with that lint will deny or forbid the unsafe external block.
+
+Within an `extern` block is zero or more declarations of external functions and/or external static values.
+An extern function is declared with a `;` instead of a function body (similar to a method of a trait).
+An extern static value is also declared with a `;` instead of an expression (similar to an associated const of a trait).
+In both cases, the actual function body or value is provided by whatever external source (which is probably not even written in Rust).
+
+When an `unsafe extern` block is used, all declarations within that `extern` block *must* have the `unsafe` or `safe` keywords as part of their signature.
+The `safe` keyword is a contextual keyword, it is currently only used within `extern` blocks.
+
+If an `extern` block is used in an older edition without the `unsafe` keyword, declarations *cannot* specify `safe` or `unsafe`.
+Code must update to `unsafe extern` style blocks if it wants to make `safe` declarations.
+
+```rust
+unsafe extern {
+    // sqrt (from libm) can be called with any `f64`
+    pub safe fn sqrt(x: f64) -> f64;
+    
+    // strlen (from libc) requires a valid pointer,
+    // so we mark it as being an unsafe fn
+    pub unsafe fn strlen(p: *const c_char) -> usize;
+
+    // this function doesn't say safe or unsafe, so it defaults to unsafe
+    pub fn free(p: *mut core::ffi::c_void);
```

Comment:
```suggestion
    // this function doesn't say safe or unsafe, so it causes a compilation error.
    // on an older edition in an `extern{}` block it would be assumed to be unsafe
    pub fn free(p: *mut core::ffi::c_void); //~ error: a function in an extern block must be marked as either `unsafe` or `safe`
```

Reply:
The proposed change here is not correct.  According to the rules that pnkfelix summarized:

> - An item without a `safe` nor `unsafe` marker within an `extern` block is implicitly unsafe (i.e. this is the current Rust 2015 semantics).